openshift-client/src/main/java/io/fabric8/openshift/client/OpenshiftAdapterSupport.java

```diff
@@ -16,63 +16,68 @@
 
 public class OpenshiftAdapterSupport {
 
-    static final ConcurrentMap<URL, Boolean> IS_OPENSHIFT = new ConcurrentHashMap<>();
-    static final ConcurrentMap<URL, Boolean> USES_OPENSHIFT_APIGROUPS = new ConcurrentHashMap<>();
+  static final ConcurrentMap<URL, Boolean> IS_OPENSHIFT = new ConcurrentHashMap<>();
+  static final ConcurrentMap<URL, Boolean> USES_OPENSHIFT_APIGROUPS = new ConcurrentHashMap<>();
+  public static final String APIS = "/apis";
 
-    public Boolean isAdaptable(Client client) {
+  public Boolean isAdaptable(Client client) {
         OpenShiftConfig config = new OpenShiftConfig(client.getConfiguration());
         if (!hasCustomOpenShiftUrl(config) && !isOpenShift(client)) {
             return false;
         }
         return true;
     }
 
-    /**
-     * Check if OpenShift is available.
-     * @param client   The client.
-     * @return         True if oapi is found in the root paths.
-     */
+  /**
+   * Check if OpenShift is available.
+   * @param client   The client.
+   * @return         True if oapi is found in the root paths.
+   */
     static boolean isOpenShift(Client client) {
-        URL masterUrl = client.getMasterUrl();
-        if (IS_OPENSHIFT.containsKey(masterUrl)) {
-            return IS_OPENSHIFT.get(masterUrl);
-        } else {
-            RootPaths rootPaths = client.rootPaths();
-            if (rootPaths != null) {
-                List<String> paths = rootPaths.getPaths();
-                if (paths != null) {
-                    for (String path : paths) {
-                        // lets detect the new API Groups APIs for OpenShift
-                        if (path.endsWith(".openshift.io") || path.contains(".openshift.io/")) {
-                            USES_OPENSHIFT_APIGROUPS.putIfAbsent(masterUrl, true);
-                            IS_OPENSHIFT.putIfAbsent(masterUrl, true);
-                            return true;
-                        }
-                    }
-                    for (String path : paths) {
-                        if (java.util.Objects.equals("/oapi", path) || java.util.Objects.equals("oapi", path)) {
-                            IS_OPENSHIFT.putIfAbsent(masterUrl, true);
-                            return true;
-                        }
-                    }
-                }
-            }
+      URL masterUrl = client.getMasterUrl();
+
+      OkHttpClient httpClient = ((BaseClient)client).getHttpClient();
+      try {
+        Request.Builder requestBuilder = new Request.Builder()
+          .get()
+          .url(URLUtils.join(masterUrl.toString(), APIS));
+        Response response = httpClient.newCall(requestBuilder.build()).execute();
+        ObjectMapper objectMapper = new ObjectMapper();
+        APIGroupList apiGroupList = objectMapper.readValue(response.body().string(), APIGroupList.class);
+
+        for (APIGroup apiGroup : apiGroupList.getGroups()) {
+          if (apiGroup.getName().endsWith("openshift.io")) {
```

Comment:
@rohanKanojia - ok, based on what I see with https://github.com/fabric8io/kubernetes-client/pull/1640/files, the `apiGroup.getName()` seems like the correct method to call to get the values that could end with `openshift.io`.

I don't see the actual generated APIGroup and APIGroupList java files in the repo yet.  I'm assuming those get generated as part of either merging this PR or some subsequent task?

Unless the actually generated java/class file unveils some sort of surprise, I think this looks good.

Reply:
Hi @gabemontero and @rohanKanojia 

it looks like the code to generate them has been changed here https://github.com/fabric8io/kubernetes-client/commit/2b29f43ca2314df7c3f93424420cf170c4f5017d#diff-4e2a515b790160e4becee6719409c1d1R70 

and maybe it was in a different PR.

@rohanKanojia could you please remember to put me also on CC with @waveywaves ?
we are handing over the Jenkins plugin code base from @gabemontero